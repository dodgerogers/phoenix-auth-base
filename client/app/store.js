import { fromJS } from 'immutable';
import { compose, createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga'
import reducer from './reducer';
import logger from './lib/middleware/logger';
import rootSaga from './sagas/rootSaga';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(reducer, compose(
  applyMiddleware(sagaMiddleware, logger),
  window.devToolsExtension ? window.devToolsExtension() : f => f
));

sagaMiddleware.run(rootSaga);


export default store;
