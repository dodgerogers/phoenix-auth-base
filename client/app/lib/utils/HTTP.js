import axios from 'axios';
import normalize from 'normalize-object';
import store from '../../store';
import { currentAccessTokenValue, isTokenRefreshing } from '../../authentication/selector';
import wait from './wait';


const localhost = 'http://localhost';
export const API_BASE = process.env.NODE_ENV === 'test' ? localhost : window.location.origin;
const HTTP = axios.create({ baseURL: API_BASE });

HTTP.interceptors.request.use(requestInterceptor, function(error) {
  return Promise.reject(error);
});

HTTP.interceptors.response.use(response => { // TODO: Handle 401
  response.data = normalize(response.data);
  return response;
}, function(error) {
  return Promise.reject(error);
});

async function requestInterceptor(config) {
  return waitForAccessToken(config.url)
    .then(accessToken => transformRequest(config, accessToken));
};

async function waitForAccessToken(url) {
  try {
    if (!isTokenRefreshRequest(url) && isTokenRefreshing(store.getState())) {
      return await wait(waitForAccessToken, url);
    }
    return currentAccessTokenValue(store.getState());
  } catch(e) {
    return null;
  }
}

function isTokenRefreshRequest(url) {
  return url && url.indexOf('/oauth/token/refresh') > -1;
}

function transformRequest(config, accessToken) {
  if (accessToken) {
    config.headers.common['Authorization'] = `Bearer ${accessToken}`;
  } else {
    const { Authorization, ...headers } = config.headers.common;
    config.headers.common = headers;
  }

  config.data = normalize(config.data, 'snake');

  return config;
}

export default HTTP;
