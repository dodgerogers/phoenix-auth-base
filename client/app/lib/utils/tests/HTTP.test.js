import { fromJS } from 'immutable';
import normalize from 'normalize-object';
import HTTP from '../HTTP';
import store from '../../../store';

describe('HTTP', () => {
  const config = () => ({
    headers: {
      common: {},
    },
    data: {
      dataPayload: {
        dataKey: 'data',
        arrayKey: [
          { nestedKey: 'key' },
        ],
      },
    },
  });

  const error = () => ({
    response: {
      statusText: 'BadRequest',
      status: 400,
      data: {
        error: {
          data_key: 'Page not found',
          array_key: [
            { data_key: 'data' },
          ],
        },
      },
    },
  });

  describe('request', () => {
    it('transforms request keys to snake case', () => {
      const initialState = {
        authentication: fromJS({
          accessToken: null,
          isTokenRefreshing: false,
        })
      };

      store.getState = jest.fn(() => initialState);

      const interceptor = HTTP.interceptors.request.handlers[0];

      return interceptor.fulfilled(config()).then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    it('adds authorization header when accessToken is present', () => {
      const initialState = {
        authentication: fromJS({
          accessToken: {
            accessToken: 'token',
          },
          isTokenRefreshing: false,
        })
      };

      store.getState = jest.fn(() => initialState);

      const interceptor = HTTP.interceptors.request.handlers[0];

      return interceptor.fulfilled(config()).then(result => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('response', () => {
    it('transforms response', () => {
      const interceptor = HTTP.interceptors.response.handlers[0];

      expect(interceptor.fulfilled(config())).toMatchSnapshot();
    });

    it('transforms response error', () => {
      const interceptor = HTTP.interceptors.response.handlers[0];

      expect(interceptor.rejected(error())).rejects.toMatchSnapshot();
    });
  });
});
