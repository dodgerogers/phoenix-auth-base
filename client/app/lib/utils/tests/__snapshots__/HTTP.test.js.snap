// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTTP request adds authorization header when accessToken is present 1`] = `
Object {
  "data": Object {
    "data_payload": Object {
      "array_key": Array [
        Object {
          "nested_key": "key",
        },
      ],
      "data_key": "data",
    },
  },
  "headers": Object {
    "common": Object {
      "Authorization": "Bearer token",
    },
  },
}
`;

exports[`HTTP request transforms request keys to snake case 1`] = `
Object {
  "data": Object {
    "data_payload": Object {
      "array_key": Array [
        Object {
          "nested_key": "key",
        },
      ],
      "data_key": "data",
    },
  },
  "headers": Object {
    "common": Object {},
  },
}
`;

exports[`HTTP response transforms response 1`] = `
Object {
  "data": Object {
    "dataPayload": Object {
      "arrayKey": Array [
        Object {
          "nestedKey": "key",
        },
      ],
      "dataKey": "data",
    },
  },
  "headers": Object {
    "common": Object {},
  },
}
`;

exports[`HTTP response transforms response error 1`] = `
Object {
  "response": Object {
    "data": Object {
      "error": Object {
        "arrayKey": Array [
          Object {
            "dataKey": "data",
          },
        ],
        "dataKey": "Page not found",
      },
    },
    "status": 400,
    "statusText": "BadRequest",
  },
}
`;
